src/v/coproc/router.cc

```diff
@@ -46,12 +47,43 @@ router::extract_offset(model::record_batch_reader reader) {
 router::router(ss::socket_address addr, ss::sharded<storage::api>& api)
   : _api(api)
   , _jitter(std::chrono::milliseconds(10))
+  , _rsm(api, _sources)
   , _transport(
       {rpc::transport_configuration{
         .server_addr = addr, .credentials = nullptr}},
       rpc::make_exponential_backoff_policy<rpc::clock_type>(
         std::chrono::seconds(1), std::chrono::seconds(10))) {}
 
+ss::future<errc> router::add_source(
+  script_id id, const model::topic_namespace& tn, topic_ingestion_policy p) {
+    if (_gate.is_closed()) {
+        return ss::make_exception_future<errc>(std::runtime_error(fmt_with_ctx(
+          fmt::format,
+          "Attempted to add a source after the routers gate has been "
+          "closed")));
+    }
+    return _rsm.add_source(id, tn, p);
+}
+
+ss::future<bool> router::remove_source(script_id sid) {
+    if (_gate.is_closed()) {
+        return ss::make_exception_future<bool>(std::runtime_error(fmt_with_ctx(
+          fmt::format,
+          "Attempted to remove a source after the routers gate has been "
+          "closed")));
+    }
+    return _rsm.remove_source(sid);
+}
+
+ss::future<> router::stop() {
+    _loop_timer.cancel();
+    _abort_source.request_abort();
+    return _gate.close().then([this] {
+        _rsm.cancel_pending_updates();
+        return _transport.stop();
```

Comment:
can transport throw?

Reply:
wasn't sure, i checked around looks like eventually shutdown() in base_transport is called which is marked noexcept. Can someone confirm?